src/Common/src/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -393,6 +396,181 @@ private static ComputedInstanceFieldLayout ComputeSequentialFieldLayout(Metadata
+
+                if (field.FieldType.IsGCPointer)
+                {
+                    if (!instanceGCPointerFieldsBySize.TryGetValue(fieldSizeAsInt, out List<FieldDesc> instanceFieldsWithMatchingSize))
+                    {
+                        instanceFieldsWithMatchingSize = new List<FieldDesc>();
+                        instanceGCPointerFieldsBySize[fieldSizeAsInt] = instanceFieldsWithMatchingSize;
+                    }
+                    instanceFieldsWithMatchingSize.Add(field);
+                }
+                else
+                {
+                    if (!instanceNonGCPointerFieldsBySize.TryGetValue(fieldSizeAsInt, out List<FieldDesc> instanceFieldsWithMatchingSize))
+                    {
+                        instanceFieldsWithMatchingSize = new List<FieldDesc>();
+                        instanceNonGCPointerFieldsBySize[fieldSizeAsInt] = instanceFieldsWithMatchingSize;
+                    }
+                    instanceFieldsWithMatchingSize.Add(field);
+                }
+            }
+
+            // Align the field offset to the required alignment of an int: 4
+            // Not sure why that is, but a .NET Core 2.1 app seems to align the subclass's fields automatically to this boundary
+            cumulativeInstanceFieldPos = LayoutInt.AlignUp(cumulativeInstanceFieldPos, new LayoutInt(4));
+
+            // First, place small fields immediately after the parent field bytes if there are a number of field bytes that are not aligned
+            // GC pointer fields and value class fields are not considered for this optimization
+            // Place the fields in order, starting with the largest size.
+            int parentByteOffsetModulo = cumulativeInstanceFieldPos.AsInt % type.Context.Target.PointerSize;
+            if (parentByteOffsetModulo != 0)
+            {
+                int alignmentBytesRemaining = type.BaseType.InstanceByteCount.AsInt - cumulativeInstanceFieldPos.AsInt;
+                bool continueProcessingInstanceFields = true;
+
+                for (int i = maxInstanceFieldSize; continueProcessingInstanceFields && i > 0; i /= 2)
+                {
+                    // Continue to next size if the current size will absolutely not fit into the remaining space
+                    if (i > alignmentBytesRemaining)
+                        continue;
+
+                    if (alignmentBytesRemaining == 0)
+                    {
+                        break;
+                    }
+
+                    Debug.Assert(alignmentBytesRemaining > 0);
+
+                    // Iterate over the variable of size i if we have stored any
+                    if (instanceNonGCPointerFieldsBySize.TryGetValue(i, out List<FieldDesc> instanceNonGCPointerFields) && instanceNonGCPointerFields.Count > 0)
+                    {
+                        while (instanceNonGCPointerFields.Count > 0)
```

Comment:
Have you got any idea what order of magnitude this count typically is? I mean, the below statement

instanceNonGCPointerFields.RemoveAt(0)

is basically O(n) in the number of list elements -  i.o.w. getting through the entire list is a O(n^2) operation. Are we confident we can ignore this, at least for now?

Reply:
Do you know of a faster operation to remove the first element? I'd be happy to use that instead of RemoveAt. Wrt ignoring this poor performance loop/method, I think we may be able to because, from my understanding, the worst case scenario would be us placing 4 bools and then we would run out of space and exit this loop.

Reply:
If the only operations on each `instanceNonGCPointerFields` list is adding to the end and drawing off index 0, use a `Queue<FieldDesc>`?

Reply:
You're absolutely right. I'll go ahead and make that change.